### ISSUE TITLE
columnar mesh definition

### ISSUE THREAD

**Issue by ecoon** [MeshToolkit/MSTK#34]
@raovgarimella  @ekikinzon The definition of what a columnar mesh is seems to get used in a few places, but to be inconsistent across those places.  In particular, 

MSTK::FixColumnPartitions() defines a column via:

```
for each region in mesh:
    top_face = argmax( (face_centroid - cell_centroid) * z_hat for f for f in faces(region) )
    if num_regions(top_face) == 1:
        # top_face is in the surface!
        # iterate through the column by finding the "bottom face" of the region
        region = regions(top_face)[0]
        bottom_face = find_first(f for f in faces(region) if nodes(f) intersect nodes(top_face) == 0 )
```

Alternatively, Mesh::build_columns_() defines a column via:

```
for each face in mesh:
    if num_regions(face) == 1 && face_normal(face) * -z_hat > 0 && (face_centroid - cell_centroid) * -z_hat > 0:
        # face is part of the bottom!
        # iterate through the column by finding the top face
        region = regions(face)[0]
        top_face = argmax( face_normal(face) * z_hat for f in faces(region) )
```

These really aren't consistent, and there are issues with both of these definitions.
1. For the first algorithm, finding the "top" face may not work if the face_centroid of one of the sides of the cell is above the face_centroid of the true top face.  This can happen in very steep cells.
2. For the first algorithm, finding the "opposing" face may not work if the region is degenerate, as then the "top" and "bottom" faces share nodes.
3. For the second algorithm, finding the "bottom" face may not work for the same reason as 1.
4. For the second algorithm, finding the "opposing" face should work, but face_normal is a little wonky on a degenerate "side" face.  As long as the normal of a degenerate face is defined reasonably, this is ok.

I would like to:
1. make these more robust
2. make these consistent EVERYWHERE the concept of a column is used.

To do this I think it is important to agree on the definition of a columnar mesh.  I believe a "columnar mesh" should be defined as: "a mesh formed by taking a 'proper surface mesh' and extruding it in the vertical by changing z-coordinates of nodes only."  A "proper surface mesh" can be mapped 1-1 to R2, i.e. all nodes are (x,y,f(x,y)) where f is a function, and has no degenerecies.

NOTE: Following the logic of this definition, a columnar mesh's faces can be partitioned into "horizontal faces", i.e. faces with arbitrary number of nodes that have normals with nonzero z-component (assumed abs(normal \* z_hat) > eps) and "side faces" which have exactly 4 nodes (as long as all edges in the surface mesh have 2 nodes and faces are never combined on extrusion) and have normal \* z_hat = 0.  Every columnar region has exactly two "horizontal faces."

NOTE: This also implies that the face_centroid of "horizontal" faces MUST be directly above the cell_centroid, i.e. norm(face_centroid_XY - cell_centroid_XY) < eps, where centroids are defined as the mean of x,y,z coordinates of all nodes and _XY indicates projection into the XY plane.  Furthermore, all "side" faces MUST have norm(face_centroid_XY - cell_centroid_XY) > eps.

Also, we need to allow degenerate regions, including those that have zero volume.  Both of the above notes are true even in the case of a degenerate region with zero volume, as long as the surface mesh is proper and has no degenerecies.  The hard part of a degenerate region is determining which of the two horizontal faces is "top" and which is "bottom".  To do this, we must iterate through neighbors until we find a non-degenerate region.

So, I would like to change this code, EVERYWHERE, to use the following algorithm for finding columns:

```
def is_side(region, face):
    return norm(face_centroid_XY - cell_centroid_XY) > eps

def find_other_horiz_face(region, face):
    return find_first(f for f in faces(region) if f != face and !is_side(region, f))

def find_other_region(face, region):
    return find_first(r for r in regions(face) if r != region)

# find the top face
for each face in mesh:
    is_column_top = False
    if num_regions(face) == 1 && !is_side(regions(face)[0], face):
        # since this is both a boundary region and a top/bottom, 
        # this must be either the top or bottom of the ENTIRE column of regions.
        if face_centroid[2] > cell_centroid(region)[2] + eps:
            # now we know this is the TOP
            is_column_top = True
        else if abs(face_centroid[2] - cell_centroid[2]) < eps:
            # degenerate, keep looking in the OTHER direction until we find a non-degenerate region
            other_horiz_face = find_other_horiz_face(region, face)
            other_region = find_other_region(other_horiz_face, region)
            while ( abs(other_horiz_face_centroid[2] - other_region_centroid[2]) < eps ):
                # still degenerate
                other_horiz_face = find_other_horiz_face(other_region, other_horiz_face)
                other_region = find_other_region(other_horiz_face, other_region)
            if other_horiz_face_centroid[2] < other_region_centroid[2] - eps:
                # other face is below the region, so this face is the top
                is_column_top = True
    if !is_column_top:
        continue

    # now iterate and find the rest of the column
    other_horiz_face = find_other_horiz_face(region, face)
    ....
```

Does this make sense?
Are there any other places than Amanzi::Mesh::build_columns() and MSTK::FixColumnPartitions() that use this logic?

**Comment by ecoon** [comment]
@ekikinzon  pointed out that the cell_centroid will not be below the face_centroid in a partially degenerate cell.

**Comment by raovgarimella** [comment]
Yep, You can get a false positive from a "horizontal" face in the "is_side" test.

**Comment by raovgarimella** [comment]
Addressed by PR #51 I think
